react/index.js

```diff
@@ -33,8 +39,97 @@ export class MiniCart extends Component {
 
   state = {
     openContent: false,
+    updatingOrderForm: false,
   }
 
+  debounce = null
+
+  async componentDidUpdate(prevProps, prevState) {
+    const serverItems = path(['data', 'orderForm', 'items'], this.props)
+    const clientItems = path(['linkState', 'minicartItems'], this.props)
+
+    if (serverItems && clientItems) {
+      const prevClientItems = path(['linkState', 'minicartItems'], prevProps)
+
+      if (!equals(prevClientItems, clientItems)) {
+        if (this.debounce) {
+          clearTimeout(this.debounce)
+        }
+        this.debounce = setTimeout(
+          () => this.handleItemsDifference({ clientItems }),
+          2e3
+        )
+      } else if (serverItems.length && !clientItems.length) {
+        return this.fillClientMinicart(serverItems)
+      } else if (prevState.updatingOrderForm && !this.state.updatingOrderForm) {
+        const {
+          data: {
+            orderForm: { items },
+          },
+        } = await this.props.data.refetch()
+        return this.fillClientMinicart(items)
+      }
+    }
+  }
+
+  handleItemsDifference = async ({ clientItems }) => {
+    const items = clientItems
+      .map(({ id, index, quantity, seller }) => ({
+        id,
+        index,
+        quantity,
+        seller,
+      }))
+      .filter(
+        ({ id, index, quantity, seller }) =>
+          !isNil(id) && !isNil(index) && !isNil(quantity) && !isNil(seller)
+      )
+    if (!items.length) return
+
+    this.setState({ updatingOrderForm: true })
+    try {
+      await this.addItems(items)
```

Comment:
If you are already filtering the items here, couldn't you make this distinction here and pass the right items to these helper functions. If I pass some items to and `addItem` function, I would hope for those items to get added. Same for `updateItem`. If not, an upsert

Reply:
I'm not already filtering the items, here. If you see these helper functions, you will know that some other things are happening and if I do it in the same `handleItemsDifference` function, it will grow a lot.